src/order/grade.lean

```diff
@@ -0,0 +1,238 @@
+
+end grade_min_order
+
+section grade_max_order
+variables (𝕆) [grade_max_order 𝕆 α] {a : α}
+
+protected lemma is_max.grade (h : is_max a) : is_max (grade 𝕆 a) := grade_max_order.is_max_grade h
+
+variables {𝕆}
+
+@[simp] lemma is_max_grade_iff : is_max (grade 𝕆 a) ↔ is_max a :=
+⟨grade_strict_mono.is_max_of_apply, is_max.grade _⟩
+
+end grade_max_order
+end preorder
+
+lemma grade_mono [partial_order α] [grade_order 𝕆 α] : monotone (grade 𝕆 : α → 𝕆) :=
+grade_strict_mono.monotone
+
+section linear_order
+variables [linear_order α] [grade_order 𝕆 α] {a b : α}
+
+lemma grade_injective : function.injective (grade 𝕆 : α → 𝕆) := grade_strict_mono.injective
+@[simp] lemma grade_le_grade_iff : grade 𝕆 a ≤ grade 𝕆 b ↔ a ≤ b := grade_strict_mono.le_iff_le
+@[simp] lemma grade_lt_grade_iff : grade 𝕆 a < grade 𝕆 b ↔ a < b := grade_strict_mono.lt_iff_lt
+@[simp] lemma grade_eq_grade_iff : grade 𝕆 a = grade 𝕆 b ↔ a = b := grade_injective.eq_iff
+lemma grade_ne_grade_iff : grade 𝕆 a ≠ grade 𝕆 b ↔ a ≠ b := grade_injective.ne_iff
+
+lemma grade_covby_grade_iff : grade 𝕆 a ⋖ grade 𝕆 b ↔ a ⋖ b :=
+(covby_iff_lt_covby_grade.trans $ and_iff_right_of_imp $ λ h, grade_lt_grade_iff.1 h.1).symm
+
+end linear_order
+end preorder
+
+section partial_order
+variables [partial_order 𝕆] [preorder α]
+
+@[simp] lemma grade_bot [order_bot 𝕆] [order_bot α] [grade_min_order 𝕆 α] : grade 𝕆 (⊥ : α) = ⊥ :=
+(is_min_bot.grade _).eq_bot
+
+@[simp] lemma grade_top [order_top 𝕆] [order_top α] [grade_max_order 𝕆 α] : grade 𝕆 (⊤ : α) = ⊤ :=
+(is_max_top.grade _).eq_top
+
+end partial_order
+
+/-! ### Instances -/
+
+variables [preorder 𝕆] [preorder α] [preorder β]
+
+instance preorder.to_grade_bounded_order : grade_bounded_order α α :=
+{ grade := id,
+  is_min_grade := λ _, id,
+  is_max_grade := λ _, id,
+  grade_strict_mono := strict_mono_id,
+  covby_grade := λ a b, id }
+
+@[simp] lemma grade_self (a : α) : grade α a = a := rfl
+
+/-! #### Dual -/
+
+instance [grade_order 𝕆 α] : grade_order (order_dual 𝕆) (order_dual α) :=
+{ grade := to_dual ∘ grade 𝕆 ∘ of_dual,
+  grade_strict_mono := grade_strict_mono.dual,
+  covby_grade := λ a b h, (h.of_dual.grade _).to_dual }
+
+instance [grade_max_order 𝕆 α] : grade_min_order (order_dual 𝕆) (order_dual α) :=
+{ is_min_grade := λ _, is_max.grade _,
+  ..order_dual.grade_order }
+
+instance [grade_min_order 𝕆 α] : grade_max_order (order_dual 𝕆) (order_dual α) :=
+{ is_max_grade := λ _, is_min.grade _,
+  ..order_dual.grade_order }
+
+instance [grade_bounded_order 𝕆 α] : grade_bounded_order (order_dual 𝕆) (order_dual α) :=
+{ ..order_dual.grade_min_order, ..order_dual.grade_max_order }
+
+@[simp] lemma grade_to_dual [grade_order 𝕆 α] (a : α) :
+  grade (order_dual 𝕆) (to_dual a) = to_dual (grade 𝕆 a) := rfl
+@[simp] lemma grade_of_dual [grade_order 𝕆 α] (a : order_dual α) :
+  grade 𝕆 (of_dual a) = of_dual (grade (order_dual 𝕆) a) := rfl
+
+/-! #### `fin n`-graded to `ℕ`-graded to `ℤ`-graded -/
```

Comment:
Do these all generalise to defs being lifted along general embeddings rather than just `fin n -> ℕ`?

Reply:
No, `grade_min_order` requires preserving `⊥` (and `grade_max_order` requires preserving `⊤`), but even for `grade_order` order embeddings aren't enough. What we need are order embeddings which preserve the covering relation. I can state it as such.

Reply:
Yeah I think that could be helpful - if nothing else it removes a bit of duplication among these defs